crates/connector_proxy/src/connector_runner.rs

```diff
@@ -75,14 +80,68 @@ pub async fn run_airbyte_source_connector(
         NetworkTunnelCaptureInterceptor::adapt_request_stream(op, request_stream())?,
     )?;
 
-    let adapted_response_stream = NetworkTunnelCaptureInterceptor::adapt_response_stream(
-        op,
-        airbyte_interceptor.adapt_response_stream(op, response_stream(child_stdout))?,
-    )?;
+    let res_stream =
+        airbyte_interceptor.adapt_response_stream(op, response_stream(child_stdout))?;
+
+    // Keep track of whether we did send a Driver Checkpoint as the final message of the response stream
+    // See the comment of the block below for why this is necessary
+    let (tp_sender, tp_receiver) = oneshot::channel::<bool>();
+    let res_stream = if *op == FlowCaptureOperation::Pull {
+        Box::pin(stream::try_unfold(
+            (false, res_stream, tp_sender),
+            |(transaction_pending, mut stream, sender)| async move {
+                let (message, raw) = match stream.next().await {
+                    Some(bytes) => {
+                        let bytes = bytes?;
+                        let mut buf = &bytes[..];
+                        let msg = decode_message::<PullResponse, _>(&mut buf)
+                            .await
+                            .unwrap()
```

Comment:
This could do with a comment saying why the decoding is infallible, assuming that's the case.